### ISSUE TITLE
HTML/Twig field doesn't save rendered HTML/Twig in submission

### ISSUE THREAD

**Issue by jakobhaerter** [verbb/formie#2182]
### Describe the bug

We're using the HTML/Twig field to display simple text or components like a list of inside a form. This works great and we can even use our predefined Twig includes (purifier disabled).

We expected that Formie saves the rendered HTML of HTML/Twig fields in each submission. However, both submission and field preview just output the raw and non-escaped value of each field. HTML is being rendered by the browser, with Twig code inside it. Which in itself is a bit inconsistent.

If this is expected behaviour, what can we do to include – for example – a list of downloads?

<img width="359" alt="image" src="https://github.com/user-attachments/assets/010a2415-02d8-4d9f-8599-bcf41f97b83a">

### Steps to reproduce

1. Create a form with formie
2. Add a HTML field and add HTML and Twig code
3. Launch the form and create a submission
4. View the submission in the CP

### Form settings

- Multi-page form: Yes
- Submission Method: Page Reload
- Client-side Validation: Yes
- Custom Form Templates: No

### Craft CMS version

5.5.3

### Plugin version

3.0.12

### Multi-site?

Yes

**Comment by engram-design** [comment]
So HTML fields are considered cosmetic and therefore don't have a value to be saved against a submission.

We could consider serializing the content of this field, but that seems like overhead, saving potentially static HTML for this field against a submission. I can see how that'd be useful in your context, but for others, it's not really necessary and just adds to database bloat.

For now, I'd say the only route is a custom field where you handle everything yourself.

**Comment by jakobhaerter** [comment]
@engram-design Thanks for your quick reply, Josh! 

That's okay, maybe we're solving this via a hook. Still, having HTML and raw Twig code inside the submission doesn't look good and is especially confusing for our clients. It would be great if you could address this in a future release!

**Comment by engram-design** [comment]
One option is to add an include to your site files, rather than include the raw Twig template, which would keep it away from "client eyes". `{% include 'forms/my-include' %}` which would contain your Twig that you'd normally include in the HTML field. That'd be if you don't want client to be able to access that sort of thing.

There's other improvements to come with the HTML field though (https://github.com/verbb/formie/issues/1709, https://github.com/verbb/formie/issues/1028).

**Comment by MoritzLost** [comment]
@engram-design I think the main issue is that the submission includes HTML fields and outputs them as raw HTML, even if they include Twig code. This never makes sense. Using an include is marginally easier on the eyes, but the output is still illegible and useless to clients. Take this HTML field (random example):

![Screenshot 2024-12-12 at 10 21 54](https://github.com/user-attachments/assets/922de456-bcbd-4690-955d-26de51dc2524)

This looks like this in the submission. The field is output as raw HTML, which means the HTML tags are rendered, but the Twig is displayed as plain text:

![Screenshot 2024-12-12 at 10 21 47](https://github.com/user-attachments/assets/58b3e6cb-41c1-4a68-b972-2cf941e6ee03)

This never makes sense. I see two options:

- Render the template at the time of submission, save the rendered HTML and display that in the submission. This means the submission would be a sort of record of what the user saw in the form (might be relevant for some use-cases). Though still not ideal since the CP will be missing the frontend styles, so it will be unstyled or partially styled HTML.
- Don't show HTML fields in the submission at all, or only show a placeholder?

**Comment by engram-design** [comment]
@MoritzLost Ah, of course - I see what you mean. Using Twig in the HTML field has always been a _slight_ stop-gap before we add full support for it, or rich text in some form (see other linked issues). While it's indeed possible to include Twig in the field (as per the field instructions), we've always been mindful of security issues regarding that.

I'm probably going to lean towards not showing HTML (or cosmetic) fields in submissions at all. However, I can see how people might like to see Heading fields in a submission.

I'd _really_ not like to record the rendered content of a HTML field if we can get away with it. I've seen people output quite a lot of HTML in these fields, which leads to a very large payload stored in a submission, leading to a large database size. It's also a security consideration.

For now, I think I'll add a field setting just for the HTML field whether to show it in submissions, like we have for email notifications.

**Comment by MoritzLost** [comment]
@engram-design Yeah, I understand your reservations. I think not showing HTML fields at all in submissions would be a good middle-ground. You could still show Heading fields as those have a defined purpose and a sensible way to display them in the submissions view. But for HTML fields, showing those in a submission will rarely make sense. It will either include Twig, in which case the content will be completely illegible to authors. Or it will just be HTML, which still will be displayed without the frontend styling, so you will get unstyled HTML.

Showing the HTML field only makes sense if you just use it just for an additional paragraph with explanations or something like that. And it would still be reasonable to not show this in the submission. Or have a setting to toggle this behaviour, as you said.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.5](https://github.com/verbb/formie/releases/tag/4.0.0-beta.5).
